### DISCUSSION TITLE
Question on keyId

### DISCUSSION THREAD

**Discussion by jj10133** [ep2p/kademlia-api#64] [Q&A] [answered]
Hi @sepgh ping after year... I will be honest I want to make P2P carpool service for my city, I have really been digging in P2P. Want your opinion on this, I was thinking of Uber H3 cell ID or parent cell ID as key for a driver within the DHT, issue is that this would be updating every 5 or 4 seconds so I was thinking of updating the routing table with nearby cells (will get cellID by H3), is this great idea? 

Another issue could be that I was thinking of grouping drivers using same cellID because it might a case, and that's where I was thinking of parent cell ID, i don't know if you know H3 but it meant for indexing but I'm completely using different.

I know that we take the xor to find the nearest node but just trying to use in this place?

Give me your opinions on this

**Comment by sepgh** [comment]
Hi @warmachinesocial 

I am not familiar with H3 cell ID system so it'll take a while till I do a little research on it. Meanwhile, would you please explain to me some of the decentralized requirements of the carpool, or some of the business requirements/logic, so I can get more familiar with the problem?

-If sharing that information here is not suitable for you let me know so we can have a private chat on Discord-

**Comment by sepgh** [comment]
After taking a brief look at H3 cell ID I have to say I really doubt it's a good idea to use it as a NodeId (or within the algorithm that generates a node id).

That being said, I still don't know your exact requirements. Since your last question was about NAT holepunching I think you are considering that your nodes are either cars (if they are that advanced) or drivers (mobile phones). In both of these cases I think generating a unique ID based on drivers or cars licenses (and maybe additional private public key if needed) would make more sense as a node ID.
Constant change in node ids would basically result in loss of data. Beside, if this system needs to be truly decentralized, proof of ownership may be required within nodes. EX: a driver should not be able able to fake the location of a car that they are not driving... (I could be off track here?)

Now for the DHT Repository Key (they key/value) that is being stored, if the point is tracking drivers and cars locations, then the H3 would begin to make sense I guess.

I hope this helps. I am looking forward to know more about the requirements.

**Comment by jj10133** [comment]
Yes, I was confused on NodeId (still think i'm not getting this). 

At the moment, less cars that's advanced so relying on drivers mobile phones. Mostly wanted for tracking locations, find the nearest drivers on client who want to book a ride

If a rider requests for a ride, could lookup the dht and retrieve the nearest driver and updating key idea was basically to match the dht key/value and nodeid (however I don't know this part for routing).

Ex -> if dht keyspace made from cellid using h3, so get rider's location and get cellid lookup the dht to see drivers nearby, if not found get surrouding cellids from rider location and lookup the dht. Same time drivers update the dht only when their cell id changes which is due to location and insert location to surrounding cellid within the dht.

DHT keys are cellID from H3 -> Update only the surround IDs

**Comment by sepgh** [comment, marked as answer]
> Yes, I was confused on NodeId (still think i'm not getting this).

The node ID is what manages the routing table and availability of the nodes within the DHT. It should not change as it is used when we are determining which node should store a DHT key/value by measuring the distance between the key and the node.


Your example makes sense to me, as long as person at points A and B can in some way have the same H3 Cell Index.
![image](https://github.com/user-attachments/assets/b301a423-5521-4372-bebc-fb751bdf80d8)

If that's true, the general idea (Despite who runs the network, since I think that may become the problem would be like this -and to agree with your idea-):


**Drivers**:

Drivers have a unique ID (which could be generated using their's and the car license for example) and as they drive around they have to keep updating the DHT by sending _a message_ to DHT that includes:

A) their ID
B) the H3 cell ID
C) their actual location (optional)
D) if they are available to accept rides
E) timestamp
F) the signature for the message


Additionally, they should also publish the same message for multiple H3 index resolutions perhaps (Ex: all the colors like below)
![image](https://github.com/user-attachments/assets/8d0f5770-dc3f-461f-9c27-2338f5bf03b9)

Within the DHT, the keys would be `H3 Cell ID` and the values could be a list of objects with drivers and their availability information.

Drivers should also send updates to DHT when they exit a H3 cell.

**Riders:**

As you mentioned, when a rider is looking for a ride, first the smallest H3 Cell should be looked up in the DHT, then it should keep growing the resolution of nodes (using a standard that matches between riders and drivers) and look up for larger areas to find drivers that are perhaps a bit more far.


**Nodes:**

As mentioned above, within the DHT, the keys would be `H3 Cell ID` and the values could be a list of objects with drivers and their availability information. When a driver exits a cell, it should be removed from the list of drivers in that cell.

Since only a single node handles a unique cell, `add`/`remove` calls to the list of drivers would also be handled by the same node.


**Riding requests**

There is a requirement of being able to ask a driver for a ride to a destination, get driver's response and perhaps see the approximate time of driver's arrival. That'd be another problem to work on.


**Network:**

Now this is the hard part. Your nodes could indeed be your drivers phones. Here are some of the problems I can imagine for such network:

- **Unavailability**: a driver could easily lose their signal. this results in a chunk of the DHT being lost. note that this chunk doesn't only include status of the current driver (whos network is unavailable), but removes multiple H3 cells from the network. We can further discuss possible solutions to this if you want.
- Additionally, that `timestamp` could be used to know if a driver device is unavailable and hasn't sent an update to DHT. Of course relying on the timestamp of the message itself makes no sense, but the timestamp of when it was stored in DHT.
- **Sharding**: this is about becoming unavailable and becoming available again. When a node leaves or lose connection to the network, other nodes would be selected to store DHT data that was previously handled by the lost node. This duplicates the H3 cells within the network. Depending on your solution to availability this may or may not be important.
- **Bootstrapping**: you need a protocol so nodes can whisper to and find each other in order to create the network. This could simply be a small centralized server. Additionally, in order for a rider to directly access the network and the DHT, they should also be able to find at least a single node. Perhaps you could instantiate the network with few nodes (drivers) that never move and have a stable address (along the internet)?
- **Driver and rider proofs**: how can the network be resilient to spam drivers and ride requests? I assume you need a strong enough verification over drivers at least.


let me know what you think

**Comment by jj10133** [comment]
I think mostly you got what I was trying to say...

For riding requests -> ETA (centrallized entity to get everything as such) but for response I was like if found the nearest driver we would get the IP? We forgot to store that in? Now I don't that's a bad idea? 

Question on DHT -> It's going to distributed so it's going to be in memory right? on unavailability haven't thought much on it but you're right 

Drivers -> proof is what I have to do causing I'm running monolith and city regulations but need to think on client not spamming. Have some idea on it but referring to Hyperswarm (https://github.com/holepunchto/hyperswarm), plus has concepts on firewall for the node just trying to understand it, sorry for the late reply

**Comment by sepgh** [comment]
> For riding requests -> ETA (centrallized entity to get everything as such) but for response I was like if found the nearest driver we would get the IP? We forgot to store that in? Now I don't that's a bad idea?

It depends on many factors, but I want to mention that if you have a p2p network bootstrapped for Kademlia DHT you don't necessarily need driver's IP address to be able to communicate to it. Knowing any node in the network should be enough to transfer a message to another node. (This is like an extension on top of kademlia... in case of this library, you'd need to create new message handlers and message types).

So you could create a message type which gets delivered to specific driver using driver ID, which includes a ride-request.

> Question on DHT -> It's going to distributed so it's going to be in memory right?

With phones the phrase memory is often misinterpreted, but it can be stored on RAM or "Disk (which phones don't actually have but you know what I mean)". The availability issue would still persist despite that. As a matter of fact, using a DB like SQlite may actually be more helpful, depending on how big your network gets.


Regarding spams, there are many ways to fight it. I just wanted to mention it just in case.

**Comment by sepgh** [comment]
Closing this issue since:

A) This is just additional information and questions regarding Kademlia protocol itself and not this library (which i am fine with btw)
B) Inactivity

Converting this to a "discussion" since I am not counting this as an "Open Issue".
